text/0000-placement-by-return.md

```diff
@@ -0,0 +1,1110 @@
+* You can't return bare unsized types. [RFC-1909](https://github.com/rust-lang/rfcs/blob/master/text/1909-unsized-rvalues.md) allows you to create them locally, and pass them as arguments, but not return them.
+
+As far as existing emplacement proposals go, this one was written with the following requirements in mind:
+
+* **It needs to be possible to wrap it in a safe API.** Safe API examples are given for built-in data structures, including a full sketch of the implementation for Box, including exception safety.
+* **It needs to support already-idiomatic constructors like `fn new() -> GiantStruct { GiantStruct { ... } }`** Since this proposal is defined in terms of Guaranteed Copy Elision, this is a gimme.
+* **It needs to be possible to in-place populate data structures that cannot be written using a single literal expression.** The `write_return_with` intrinsic suggested in this proposal allows this to be done in an unsafe way. Sketches for APIs built on top of them are also given in the [future-possibilities] section.
+* **It needs to avoid adding egregious overhead in cases where the values being populated are small (in other words, if the value being initialized is the size of a pointer or smaller, it needs to be possible for the compiler to optimize away the outptr).** Since this proposal is written in terms of Guaranteed Copy Elision, this is a gimme. The exception of the "weird bypass functions" `read_return_with` and `write_return_with` may seem to break this; see the [example desugarings here](#How-do-the-return-slot-functions-work-when-the-copy-is-not-actually-elided) for info on how these functions work when the copy is not actually elided.
+* **It needs to solve most of the listed problems with the old proposals.** Since this one actually goes the distance and defines when copy elision will kick in, it fixes the biggest problems that the old `box` literal system had. It is also written in terms of present-day Rust, using `impl Trait`, and with the current direction of Rust in mind.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+If you need to allocate a very large data structure, or a DST, you should prefer using `Box::new_with` over `Box::new`. For example:
+
+```rust
+let boxed_array = Box::new_with(|| [0; 1_000_000]); // instead of Box::new([0; 1_000_000])
+let boxed_data_struct = Box::new_with(DataStruct::new); // instead of Box::new(DataStruct::new())
+```
+
+The `new_with` function will perform the allocation first, then evaluate the closure, placing the result directly within it. The `new` function, on the other hand, evaluates the argument *then* performs the allocation and copies the value into it.
```

Comment:
I don't believe this behavior is set in stone. `Box::new` could be optimized to allocate *before* evaluating the argument, then you wouldn't need any new API surface for zero-copy `Box::new`.

Reply:
Does, in general, Rust consider (not regarding how it's conveyed to LLVM) calls to the allocator a side-effect that is non-orderable with other side-effecting code? To enable these kinds of optimisations, it shouldn't, but I wonder what's the deal with user-provided allocators and such.

Reply:
No, allocator calls are not considered to have important side-effects from what I understand. LLVM is free to optimize allocations out completely, after all.

Reply:
The reordering has the potential to happen only if `Box::new` is inlined, though.

More importantly, Box is the easy case. Something like `Arc::new` probably can't be trivially reordered, because it mutates shared global state.

Reply:
There is no "shared global state" in `Arc::new`, those ref counters are still "private" before return. Perhaps `Vec::push` is better example?

But even if LLVM is free to reorder the calls to eliminate the memcpy, it is also free to *not able to* perform that optimization. I'd still prefer to have this RFC to ensure the evaluation happens in-place.

Reply:
Yeah, `Arc::new` was a bad example. I thought of `Vec::push` too, but I'm not even sure LLVM can't optimize it.

> But even if LLVM is free to reorder the calls to eliminate the memcpy, it is also free to not able to perform that optimization.

Right, that's a better argument all around.

(also this RFC enables unsized returns, which goes beyond what LLVM optimizations can do)